### PR TITLE
Replace communities_users with GroupMembership (WIP)

### PR DESCRIPTION
**Pull request by levity**
Branched from #390, which is not yet merged, hence the diff below is larger than actual size.

TODO

- [x] res.locals.membership (no longer used)
- [x] set role in db to default to 0
- [x] migrate new_post_count into settings
- [x] fix sharedCommunityMembership graphql filter
- [x] add `group_data_type` to GroupMembership? This would allow looking up the corresponding group data object in a single joined query (see `GroupMembership.findGroupData` for the problem)
- [x] fix tests
- [x] keep `Community.num_members` up to date
- [x] `group_memberships.active` needs to be set to false if the group is deactivated
- [x] make `sendEmail` and `sendPushNotifications` camelcased everywhere
- [x] make sure lastReadAt is also available as lastViewedAt for communities
- [x] solve `newPostCount` updating problem -- upgrade DB or move it out of JSON

### CODE REVIEW

File: api/models/MembershipDeprecated.js [hunk 1]
```diff
@@ -18,7 +18,7 @@ module.exports = bookshelf.Model.extend(merge({
   },
```
[the remark this answers is not in the document]
Comment (levity) [reply api/models/MembershipDeprecated.js]: This PR isn't actually ready to review yet

File: api/services/InvitationService.js [hunk 2]
```diff
@@ -155,13 +155,14 @@ module.exports = {
     }
   },
 
-  use: (userId, token, accessCode) => {
+  async use (userId, token, accessCode) {
```
Comment (razorman8669) [inline api/services/InvitationService.js]: this method used to throw a plain unwrapped error.  are you sure that it's calling function handles it correctly now with async?

File: api/models/MembershipDeprecated.js [hunk 3]
```diff
@@ -18,7 +18,7 @@ module.exports = bookshelf.Model.extend(merge({
...
```
[the remark this answers is not in the document]
Comment (robbiecarlton) [reply api/models/MembershipDeprecated.js]: We still use the class in the migrations, so we will be keeping this class around for as long as we want the migrations to be available. We don't currently have anything in place for cleaning up old migrations, but typically we leave them lying around for awhile.
At minimum, we would need them around until this is live in production and all devs have run the migrations locally.

File: api/models/User.js [hunk 4]
```diff
@@ -102,16 +96,19 @@ module.exports = bookshelf.Model.extend(merge({
     return this.hasMany(Thank)
   },
 
-  joinCommunity: function (community, role = Membership.DEFAULT_ROLE) {
-    var communityId = (typeof community === 'object' ? community.id : community)
-    return Membership.create(this.id, communityId, {role})
-    .tap(() => this.markInvitationsUsed(communityId))
+  async joinCommunity (community, role = GroupMembership.Role.DEFAULT, { transacting } = {}) {
```
[the remark this answers is not in the document]
Comment (robbiecarlton) [reply api/models/User.js]: fixed

**Pull request merged**